README.md
# Ghost Driver

Ghost Driver is a pure JavaScript implementation of the
[WebDriver Wire Protocol](http://code.google.com/p/selenium/wiki/JsonWireProtocol)
for [PhantomJS](http://phantomjs.org/).
It's a Remote WebDriver that uses PhantomJS as back-end.

**GhostDriver is designed to be integral part of PhantomJS itself, but it's developed in isolation and progress is tracked
by this Repository.**

* Current _GhostDriver_ stable version:
see [releases](https://github.com/detro/ghostdriver/releases)
* Current _PhantomJS-integrated_ version is `"1.1.0"`:
contained in PhantomJS `"1.9.7"`
* Current _PhantomJSDriver Java bindings_ stable version: see
[Maven](https://oss.sonatype.org/index.html#nexus-search;quick~phantomjsdriver)

For more info, please take a look at the [changelog](https://github.com/detro/ghostdriver/blob/master/CHANGELOG.md).

The project was created and is lead by [Ivan De Marino](https://github.com/detro).

## Setup

* Download latest stable PhantomJS from [here](http://phantomjs.org/download.html)
* Selenium version `">= 2.33.0`"

**THAT'S IT!!** Because of latest stable GhostDriver being embedded in PhantomJS,
you shouldn't need anything else to get started.

## Register GhostDriver with a Selenium Grid hub

1. Launch the grid server, which listens on 4444 by default: `java -jar /path/to/selenium-server-standalone-<SELENIUM VERSION>.jar -role hub`
2. Register with the hub: `phantomjs --webdriver=8080 --webdriver-selenium-grid-hub=http://127.0.0.1:4444`
3. Now you can use your normal webdriver client with `http://127.0.0.1:4444` and just request `browserName: phantomjs`

## (Java) Bindings

This project provides WebDriver bindings for Java under the name _PhantomJSDriver_.
[Here is the JavaDoc](https://cdn.rawgit.com/detro/ghostdriver/master/binding/java/docs/javadoc/index.html).

Bindings for other languages (C#, Python, Ruby, ...) are developed and maintained
under the same name within the [Selenium project](http://docs.seleniumhq.org/docs/) itself.

### Include Java Bindings in your Maven project

Just add the following to your `pom.xml`:

```xml
<dependency>
    <groupId>com.github.detro.ghostdriver</groupId>
    <artifactId>phantomjsdriver</artifactId>
    <version>LATEST_VERSION_HERE</version>
</dependency>
```

### Include Java Bindings in your Gradle project

Just add the following to your `build.gradle`:

```gradle
dependencies {
    ...
    testCompile "com.github.detro.ghostdriver:phantomjsdriver:LATEST_VERSION_HERE"
    ...
}
```

### Search Maven
For other options on how to include the PhantomJSDriver dependency into your project, click [here](http://search.maven.org/#search%7Cga%7C1%7Ccom.github.detro.ghostdriver).

### Alternative: how to use it via `RemoteWebDriver`

Launching PhantomJS in Remote WebDriver mode it's simple:
```bash
$ phantomjs --webdriver=PORT
```
Once started, you can use any `RemoteWebDriver` implementation to send commands to it. I advice to take a look to the
`/test` directory for examples.

## F.A.Q.

### What extra WebDriver `capabilities` GhostDriver offers?
* GhostDriver extra Capabilities
    * `phantomjs.page.settings.SETTING = VALUE` - Configure `page.settings`
    on PhantomJS internal page objects (_windows_ in WebDriver context)
    (see [reference](https://github.com/ariya/phantomjs/wiki/API-Reference-WebPage#wiki-webpage-settings))
    * `phantomjs.page.customHeaders.HEADER = VALUE` - Add extra HTTP Headers
    when loading a URL
    (see [reference](https://github.com/ariya/phantomjs/wiki/API-Reference-WebPage#wiki-webpage-customHeaders))
* PhantomJSDriver (Java-binding) Capabilities
    * `phantomjs.binary.path` - Specify path to PhantomJS executable to use
    * `phantomjs.ghostdriver.path` - Specify path to GhostDriver `main/src.js`
    script to use; allows to use a different version of GhostDriver then the one
    embed in PhantomJS
    * `phantomjs.cli.args` - Specify command line arguments to pass to the
    PhantomJS executable
    * `phantomjs.ghostdriver.cli.args` - Specify command line argument to pass to
    GhostDriver (works only in tandem with `phantomjs.ghostdriver.path`)

## Want to help? Read on!

GhostDriver pushed the evolution of PhantomJS from the start. All the features required by PhantomJS to fit GhostDriver were designed to still feel "consistent" and "at home" with PhantomJS alone.

To drive that effort, I worked on a [PhantomJS fork](https://github.com/detro/phantomjs-ghostdriver), and then
pushed changes to PhantomJS master once agreed with the rest of the team on the changes.

If you are planning to contribute, that is the PhantomJS you should use.

### Run validation tests

Here I show how to clone this repo and kick start the (Java) tests. You need
[Java SDK](http://www.oracle.com/technetwork/java/javase/downloads/index.html)
to run them (I tested it with Java 7, but should work with Java 6 too).

1. `git clone https://github.com/detro/ghostdriver.git`
2. Configure `phantomjs_exec_path` inside `ghostdriver/test/config.ini` to point at the build of PhantomJS you just did
3. `cd ghostdriver/test/java; ./gradlew test`

#### Alternative: Run GhostDriver yourself and launch tests against that instance

1. `phantomjs --webdriver=PORT`
2. Configure `driver` inside `ghostdriver/test/config.ini` to point at the URL `http://localhost:PORT`
3. `cd ghostdriver/test/java; ./gradlew test`

### Project Directory Structure

Here follows the output of the `tree -hd -L 3` command, trimmed of files and "build directories":

```bash
.
├── [ 102]  binding
│   └── [ 510]  java
│       ├── [ 204]  build
│       ├── [ 136]  gradle
│       ├── [ 884]  jars            <--- JARs containing Binding, related Source and related JavaDoc
│       └── [ 102]  src             <--- Java Binding Source
├── [ 442]  src                     <--- GhostDriver JavaScript core source
│   ├── [ 306]  request_handlers    <--- JavaScript "classes/functions" that handle HTTP Requests
│   └── [ 204]  third_party         <--- Third party/utility code
│       └── [2.0K]  webdriver-atoms <--- WebDriver Atoms, automatically imported from the Selenium project
├── [ 204]  test
│   ├── [ 476]  java                <--- Java Tests
│   │   ├── [ 136]  gradle
│   │   ├── [ 136]  out
│   │   └── [ 102]  src
│   ├── [ 238]  python              <--- Python Tests
│   │   └── [ 102]  utils
│   └── [ 340]  testcase-issue_240
└── [ 238]  tools                   <--- Tools (import/export)
    └── [ 136]  atoms_build_dir

20 directories
```

### WebDriver Atoms

Being GhostDriver a WebDriver implementation, it embeds the standard/default WebDriver Atoms to operate inside open
webpages. In the specific, the Atoms cover scenarios where the "native" PhantomJS `require('webpage')` don't stretch.

Documentation about how those work can be found [here](http://code.google.com/p/selenium/wiki/AutomationAtoms)
and [here](http://www.aosabook.org/en/selenium.html).

How are those Atoms making their way into GhostDriver? If you look inside the `/tools` directory you can find a bash
script: `/tools/import_atoms.sh`. That script accepts the path to a Selenium local repo, runs the
[CrazyFunBuild](http://code.google.com/p/selenium/wiki/CrazyFunBuild) to produce the compressed/minified Atoms,
grabs those and copies them over to the `/src/third_party/webdriver-atoms` directory.

The Atoms original source lives inside the Selenium repo in the subtree of `/javascript`. To understand how the build
works, you need to spend a bit of time reading about
[CrazyFunBuild](http://code.google.com/p/selenium/wiki/CrazyFunBuild): worth your time if you want to contribute to
GhostDriver (or any WebDriver, as a matter of fact).

One thing it's important to mention, is that CrazyFunBuild relies on the content of `build.desc` file to understand
what and how to build it. Those files define what exactly is built and what it depends on. In the case of the Atoms,
the word "build" means "run Google Closure Compiler over a set of files and compress functions into Atoms".
The definition of the Atoms that GhostDriver uses lives at `/tools/atoms_build_dir/build.desc`.

Let's take this small portion of our `build.desc`:
```
js_deps(name = "deps",
  srcs = "*.js",
  deps = ["//javascript/atoms:deps",
          "//javascript/webdriver/atoms:deps"])

js_fragment(name = "get_element_from_cache",
  module = "bot.inject.cache",
  function = "bot.inject.cache.getElement",
  deps = [ "//javascript/atoms:deps" ])

js_deps(name = "build_atoms",
  deps = [
    ...
    "//javascript/webdriver/atoms:execute_script",
    ...
  ]
```
The first part (`js_deps(name = "deps"...`) declares what are the dependency of this `build.desc`: with that CrazyFunBuild knows
what to build before fulfilling our build.

The second part (`js_fragment(...`) defines an Atom: the `get_element_from_cache` is going to be the name of
an Atom to build; it can be found in the module `bot.inject.cache` and is realised by the function named
`bot.inject.cache.getElement`.

The third part (`js_deps(name = "build_atoms"...`) is a list of the Atoms (either defined by something like the second
part or in one of the files we declared as dependency) that we want to build.

If you reached this stage in understanding the Atoms, you are ready to go further by yourself.

### Contributions and/or Bug Report

You can contribute by testing GhostDriver, reporting bugs and issues, or submitting Pull Requests.
Any **help is welcome**, but bear in mind the following base principles:

* Issue reporting requires a reproducible example, otherwise will most probably be **closed without warning**
* Squash your commits by theme: I prefer a clean, readable log
* Maintain consistency with the code-style you are surrounded by
* If you are going to make a big, substantial change, let's discuss it first
* I **HATE** CoffeeScript: assume I'm going to laugh off any "contribution" that contains such _aberrating crap_!
* Open Source is NOT a democracy (and I mean it!)

## License
GhostDriver is distributed under [BSD License](http://www.opensource.org/licenses/BSD-2-Clause).

## Release names
See [here](http://en.wikipedia.org/wiki/List_of_ghosts).


CHANGELOG.md
# CHANGELOG

## v1.2.0 (?-?-?) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.2.0&state=closed)) **NOT RELEASED YET**

### JavaScript Driver (Core)

### Java Binding

## v1.1.1 (2014-01-12) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.1.1&state=closed))

### JavaScript Driver (Core)
* ENHANCEMENT #170: Implement Session Isolation

## v1.1.0 (2014-01-01) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.1.0&state=closed))

### JavaScript Driver (Core)
* ENHANCEMENT: `/maximize` window will set the window size to 1336x768,
currently most common resolution online (see [statcounter](http://gs.statcounter.com/#resolution-ww-monthly-201307-201312))
* ENHANCEMENT #275: Implemented Browser and Network (HAR) Logging types
* FIXED #284: Attempt to wait for Page to Load if input causes form submit
* FIXED #291: Throw exception when attempting to set invalid timeout value
* FIXED #259: Fix issue regarding mouse clicks
* ENHANCEMENT #290: Enabled support for "Keep Alive" HTTP connections
* ENHANCEMENT #262: Allow access to PhantomJS API from WebDriver (Driver part)
* ENHANCEMENT #293: Import Selenium 2.39.0 WebDriver Atoms

### Java Binding
* MINOR #251: Minor compilation issues for Binding
* ENHANCEMENT #262: Allow access to PhantomJS API from WebDriver (Java Binding part)

## v1.0.4 (2013-07-25) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.0.4&state=closed))

### JavaScript Driver (Core)
* ENHANCEMENT: Completely refactored the way events that initiate PAGE_LOAD events are handled
* FIX #18: Now GhostDriver works when enabling `--remote-debugger-port=PORT` in PhantomJS
* FIX #180: Avoid crashes when JS execution invokes `window.close()`
* ENHANCEMENT #202: Ignore page load failures initiated by Click action (like other Drivers)
* ENHANCEMENT #208: Added the ability to shutdown logging (i.e. Log Level `OFF` or `NONE`)
* ENHANCEMENT #210: Ensuring we don't switch windows via empty string
* FIX #213: `mouseButtonDown` was broken
* FIX #215: Improve URL parsing to ensure we don't take `/status` as a URL instead of part of the protocol
* FIX #216: Fixed issue where sometimes a failed click would cause a `Parse error`
* ENHANCEMENT #228: Allow the setting of `page.customheaders` via Capabilities
* FIX #240: Stopping all ongoing HTTP requests before starting to "wait for page load" was not a good idea
* ENHANCEMENT #242: Trim URL before trying to load it
* FIX #247: Implement changes to "HTTP POST /session" as per new WireProtocol specs

### Java Binding
* ENHANCEMENT/FIX #179: Suppor for `.withLog(logfile)` option when using `PhantomJSDriverService`
* ENHANCEMENT #228: Support in the Java Bindings to set `page.customheaders` via Capabilities
* FIX 246: Avoid Cast Exception in Binding when passing `phantomjs.cli.args` through Selenium Server or Grid

### PhantomJS (code in PhantomJS master repo)
* FIX #204: PhantomJS `webserver` module should not assume UPPERCASE or lowercase for HTTP headers, as per [RFC 2616](http://www.ietf.org/rfc/rfc2616.txt) - [see also](https://github.com/ariya/phantomjs/issues/11421)

### Test
* Many more Java test to prove or disprove issues that have been submitted

## v1.0.3 (2013-03-18) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.0.3&state=closed))

### JavaScript Driver (Core)
* FIX: Rewritten code that deals with "wait for Page Load" on Clicks (based on new `page.loading` property - see below) - This closed multiple issues
* FIX: `sendKeys('\n')` should deliver an `Enter` keycode, not `Return`
* FIX: New Sessions have at least 1 Window on creation: this stops the `_cleanWindowlessSessions` code to delete brand new session that haven't yet been used
* FIX: Handle Frame-Switching when frame has no name
* FIX: Handle Frame-Switching when frame has no name and no #id
* FIX: Ensure File exists before attempting Upload
* ENHANCEMENT: Upgraded WebDriver Atoms to **Selenium 2.31.0**

### Java Binding
* FIX: Issues with "save_screenshot" method
* FIX: Java Binding should handle Proxy configuration provided as a Map in the Capabilities (like Ruby or NodeJS bindings do)
* FIX: `PhantomJSDriver.Builder.withLogFile()` is now implemented correctly and not ignored
* ENHANCEMENT: Ability to pass CLI Arguments to PhantomJS process via Capability `phantomjs.cli.args`

### PhantomJS (code in PhantomJS master repo)
* ENHANCEMENT: Added properties `page.loading` and `page.loadingProgress` to help track Page Loading


## v1.0.2 (2012-12-20) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.0.2&state=closed))

### JavaScript Driver (Core)
* FIX: Wait for potential "Page Load" when "Element.click()" is used on MORE Elements


## v1.0.1 (2012-12-15) ([issues](https://github.com/detro/ghostdriver/issues?labels=1.0.1&state=closed))

### JavaScript Driver (Core)
* ENHANCEMENT: Allow to set `page.settings` via Capabilities
* FIX: At least 1 "current window handle" should be available on a new WebDriver instance
* FIX: SwitchTo().Window() using the Window Handle
* FIX: Invalid response when checking element equality
* ENHANCEMENT: Wait for potential "Page Load" when "Element.click()" is used
* ENHANCEMENT: Command `/shutdown` arrives via HTTP HEAD in some language bindings
* ENHANCEMENT: Timeouts must be set to INFINITE by default

### PhantomJS (code in PhantomJS master repo)
* FIX: Command line argument --webdriver-selenium-grid-hub is not taken into account

### Java Binding
* BUG: PhantomJSDriverService replaced by DriverService in PhantomJSDriver constructor


## v1.0.0 (2012-11-25)
### JavaScript Driver (Core)
* Implemented all the WireProtocol commands planned for 1.0.0 ([see spreadsheet](https://docs.google.com/spreadsheet/ccc?key=0Am63grtxc7bDdGNqX1ZPX2VoZlE2ZHZhd09lNDkzbkE))

### Java Binding
* Java Binding `jar` files created in `/binding/java/jars`
* Bindings committed as _third party library_ into the Selenium project ([commit](https://code.google.com/p/selenium/source/detail?r=18187))


src/errors.js
/*
This file is part of the GhostDriver by Ivan De Marino <http://ivandemarino.me>.

Copyright (c) 2012-2014, Ivan De Marino <http://ivandemarino.me>
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

//------------------------------------------------------- Invalid Request Errors
//----- http://code.google.com/p/selenium/wiki/JsonWireProtocol#Invalid_Requests
exports.INVALID_REQ = {
    "UNKNOWN_COMMAND"               : "Unknown Command",
    "UNIMPLEMENTED_COMMAND"         : "Unimplemented Command",
    "VARIABLE_RESOURCE_NOT_FOUND"   : "Variable Resource Not Found",
    "INVALID_COMMAND_METHOD"        : "Invalid Command Method",
    "MISSING_COMMAND_PARAMETER"     : "Missing Command Parameter"
};

var _invalidReqHandle = function(res) {
    // Set the right Status Code
    switch(this.name) {
        case exports.INVALID_REQ.UNIMPLEMENTED_COMMAND:
            res.statusCode = 501;   //< 501 Not Implemented
            break;
        case exports.INVALID_REQ.INVALID_COMMAND_METHOD:
            res.statusCode = 405;   //< 405 Method Not Allowed
            break;
        case exports.INVALID_REQ.MISSING_COMMAND_PARAMETER:
            res.statusCode = 400;   //< 400 Bad Request
            break;
        default:
            res.statusCode = 404;   //< 404 Not Found
            break;
    }

    res.setHeader("Content-Type", "text/plain");
    res.writeAndClose(this.name + " - " + this.message);
};

// Invalid Request Error Handler
exports.createInvalidReqEH = function(errorName, req) {
    var e = new Error();

    e.name = errorName;
    e.message = JSON.stringify(req);
    e.handle = _invalidReqHandle;

    return e;
};
exports.handleInvalidReqEH = function(errorName, req, res) {
    exports.createInvalidReqEH(errorName, req).handle(res);
};

// Invalid Request Unknown Command Error Handler
exports.createInvalidReqUnknownCommandEH = function(req) {
    return exports.createInvalidReqEH (
        exports.INVALID_REQ.UNKNOWN_COMMAND,
        req);
};
exports.handleInvalidReqUnknownCommandEH = function(req, res) {
    exports.createInvalidReqUnknownCommandEH(req).handle(res);
};

// Invalid Request Unimplemented Command Error Handler
exports.createInvalidReqUnimplementedCommandEH = function(req) {
    return exports.createInvalidReqEH (
        exports.INVALID_REQ.UNIMPLEMENTED_COMMAND,
        req);
};
exports.handleInvalidReqUnimplementedCommandEH = function(req, res) {
    exports.createInvalidReqUnimplementedCommandEH(req).handle(res);
};

// Invalid Request Variable Resource Not Found Error Handler
exports.createInvalidReqVariableResourceNotFoundEH = function(req) {
    return exports.createInvalidReqEH (
        exports.INVALID_REQ.VARIABLE_RESOURCE_NOT_FOUND,
        req);
};
exports.handleInvalidReqVariableResourceNotFoundEH = function(req, res) {
    exports.createInvalidReqVariableResourceNotFoundEH(req).handle(res);
};

// Invalid Request Invalid Command Method Error Handler
exports.createInvalidReqInvalidCommandMethodEH = function(req) {
    return exports.createInvalidReqEH (
        exports.INVALID_REQ.INVALID_COMMAND_METHOD,
        req);
};
exports.handleInvalidReqInvalidCommandMethodEH = function(req, res) {
    exports.createInvalidReqInvalidCommandMethodEH(req).handle(res);
};

// Invalid Request Missing Command Parameter Error Handler
exports.createInvalidReqMissingCommandParameterEH = function(req) {
    return exports.createInvalidReqEH (
        exports.INVALID_REQ.MISSING_COMMAND_PARAMETER,
        req);
};
exports.handleInvalidReqMissingCommandParameterEH = function(req, res) {
    exports.createInvalidReqMissingCommandParameterEH(req).handle(res);
};

//-------------------------------------------------------- Failed Command Errors
//------ http://code.google.com/p/selenium/wiki/JsonWireProtocol#Failed_Commands
exports.FAILED_CMD_STATUS = {
    "SUCCESS"                       : "Success",
    "NO_SUCH_ELEMENT"               : "NoSuchElement",
    "NO_SUCH_FRAME"                 : "NoSuchFrame",
    "UNKNOWN_COMMAND"               : "UnknownCommand",
    "STALE_ELEMENT_REFERENCE"       : "StaleElementReference",
    "ELEMENT_NOT_VISIBLE"           : "ElementNotVisible",
    "INVALID_ELEMENT_STATE"         : "InvalidElementState",
    "UNKNOWN_ERROR"                 : "UnknownError",
    "ELEMENT_IS_NOT_SELECTABLE"     : "ElementIsNotSelectable",
    "JAVA_SCRIPT_ERROR"             : "JavaScriptError",
    "XPATH_LOOKUP_ERROR"            : "XPathLookupError",
    "TIMEOUT"                       : "Timeout",
    "NO_SUCH_WINDOW"                : "NoSuchWindow",
    "INVALID_COOKIE_DOMAIN"         : "InvalidCookieDomain",
    "UNABLE_TO_SET_COOKIE"          : "UnableToSetCookie",
    "UNEXPECTED_ALERT_OPEN"         : "UnexpectedAlertOpen",
    "NO_ALERT_OPEN_ERROR"           : "NoAlertOpenError",
    "SCRIPT_TIMEOUT"                : "ScriptTimeout",
    "INVALID_ELEMENT_COORDINATES"   : "InvalidElementCoordinates",
    "IME_NOT_AVAILABLE"             : "IMENotAvailable",
    "IME_ENGINE_ACTIVATION_FAILED"  : "IMEEngineActivationFailed",
    "INVALID_SELECTOR"              : "InvalidSelector"
};
exports.FAILED_CMD_STATUS_CODES = {
    "Success"                   : 0,
    "NoSuchElement"             : 7,
    "NoSuchFrame"               : 8,
    "UnknownCommand"            : 9,
    "StaleElementReference"     : 10,
    "ElementNotVisible"         : 11,
    "InvalidElementState"       : 12,
    "UnknownError"              : 13,
    "ElementIsNotSelectable"    : 15,
    "JavaScriptError"           : 17,
    "XPathLookupError"          : 19,
    "Timeout"                   : 21,
    "NoSuchWindow"              : 23,
    "InvalidCookieDomain"       : 24,
    "UnableToSetCookie"         : 25,
    "UnexpectedAlertOpen"       : 26,
    "NoAlertOpenError"          : 27,
    "ScriptTimeout"             : 28,
    "InvalidElementCoordinates" : 29,
    "IMENotAvailable"           : 30,
    "IMEEngineActivationFailed" : 31,
    "InvalidSelector"           : 32
};
exports.FAILED_CMD_STATUS_CODES_NAMES       = [];
exports.FAILED_CMD_STATUS_CODES_NAMES[0]    = "Success";
exports.FAILED_CMD_STATUS_CODES_NAMES[7]    = "NoSuchElement";
exports.FAILED_CMD_STATUS_CODES_NAMES[8]    = "NoSuchFrame";
exports.FAILED_CMD_STATUS_CODES_NAMES[9]    = "UnknownCommand";
exports.FAILED_CMD_STATUS_CODES_NAMES[10]   = "StaleElementReference";
exports.FAILED_CMD_STATUS_CODES_NAMES[11]   = "ElementNotVisible";
exports.FAILED_CMD_STATUS_CODES_NAMES[12]   = "InvalidElementState";
exports.FAILED_CMD_STATUS_CODES_NAMES[13]   = "UnknownError";
exports.FAILED_CMD_STATUS_CODES_NAMES[15]   = "ElementIsNotSelectable";
exports.FAILED_CMD_STATUS_CODES_NAMES[17]   = "JavaScriptError";
exports.FAILED_CMD_STATUS_CODES_NAMES[19]   = "XPathLookupError";
exports.FAILED_CMD_STATUS_CODES_NAMES[21]   = "Timeout";
exports.FAILED_CMD_STATUS_CODES_NAMES[23]   = "NoSuchWindow";
exports.FAILED_CMD_STATUS_CODES_NAMES[24]   = "InvalidCookieDomain";
exports.FAILED_CMD_STATUS_CODES_NAMES[25]   = "UnableToSetCookie";
exports.FAILED_CMD_STATUS_CODES_NAMES[26]   = "UnexpectedAlertOpen";
exports.FAILED_CMD_STATUS_CODES_NAMES[27]   = "NoAlertOpenError";
exports.FAILED_CMD_STATUS_CODES_NAMES[28]   = "ScriptTimeout";
exports.FAILED_CMD_STATUS_CODES_NAMES[29]   = "InvalidElementCoordinates";
exports.FAILED_CMD_STATUS_CODES_NAMES[30]   = "IMENotAvailable";
exports.FAILED_CMD_STATUS_CODES_NAMES[31]   = "IMEEngineActivationFailed";
exports.FAILED_CMD_STATUS_CODES_NAMES[32]   = "InvalidSelector";

var _failedCommandHandle = function(res) {
    // Generate response body
    var body = {
        "sessionId" : this.errorSessionId,
        "status" : this.errorStatusCode,
        "value" : {
            "message" : this.message,
            "screen" : this.errorScreenshot,
            "class" : this.errorClassName
        }
    };

    // Send it
    res.statusCode = 500; //< 500 Internal Server Error
    res.writeJSONAndClose(body);
};

// Failed Command Error Handler
exports.createFailedCommandEH = function(errorName, errorMsg, req, session, className) {
    var e = new Error();

    e.name = errorName;
    e.message = JSON.stringify({ "errorMessage" : errorMsg, "request" : req });
    e.errorStatusCode = exports.FAILED_CMD_STATUS_CODES[errorName] || 13; //< '13' Unkown Error
    e.errorSessionId = session.getId() || null;
    e.errorClassName = className || "unknown";
    e.errorScreenshot = (session.getCapabilities().takesScreenshot && session.getCurrentWindow() !== null) ?
        session.getCurrentWindow().renderBase64("png") : "";
    e.handle = _failedCommandHandle;

    return e;
};
exports.handleFailedCommandEH = function(errorName, errorMsg, req, res, session, className) {
    exports.createFailedCommandEH(errorName, errorMsg, req, session, className).handle(res);
};
